ledger/store/src/helpers/rocksdb/internal/map.rs

```diff
@@ -173,26 +175,33 @@ impl<
                 })
                 .collect::<Result<Vec<_>>>()?;
 
-            // Prepare operations batch for underlying database.
-            let mut batch = WriteBatch::default();
-
-            // Perform all the queued operations.
+            // Enqueue all the operations from the map in the database-wide batch.
+            let mut atomic_batch = self.database.atomic_batch.lock();
             for (raw_key, raw_value) in prepared_operations {
                 match raw_value {
-                    Some(raw_value) => batch.put(raw_key, raw_value),
-                    None => batch.delete(raw_key),
+                    Some(raw_value) => atomic_batch.put(raw_key, raw_value),
+                    None => atomic_batch.delete(raw_key),
                 };
             }
-
-            // Execute all the operations atomically.
-            self.database.rocksdb.write(batch)?;
         }
 
         // Clear the checkpoint stack.
-        *self.checkpoint.lock() = Default::default();
+        self.checkpoints.lock().clear();
         // Set the atomic batch flag to `false`.
         self.batch_in_progress.store(false, Ordering::SeqCst);
 
+        // Subtract the atomic depth index.
+        let previous_atomic_depth = self.database.atomic_depth.fetch_sub(1, Ordering::SeqCst);
+
+        // If we're at depth 0, it is the final call to `finish_atomic` and the
+        // atomic write batch can be physically executed.
+        if previous_atomic_depth == 1 {
```

Comment:
Do we need to check the edge case if `self.database.atomic_depth` is already zero? Since `fetch_sub` does a wrap if we subtract from 0

Reply:
I can add a relevant assertion; all the calls to `start_atomic` should have corresponding calls to `finish_atomic`, so the final value should always be 0.